File: leetcode/algorithms/binary-search/online-election.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Online Election"
*/

class TopVotedCandidate {
    private List<ArrayList<Integer>> leadChangeList = null;
    
    private ArrayList<Integer> getClosestLeadEntryForTime(int time){
        int low=0,high=leadChangeList.size()-1;
        ArrayList<Integer> closestEntry = null;
        while(high-low>1){
            // System.out.println(Integer.toString(low)+" "+Integer.toString(high));
            int mid = (high+low)/2;
            int midValue = leadChangeList.get(mid).get(0);
            
            if(midValue==time){
                closestEntry = leadChangeList.get(mid);
                break;
            } else if(midValue<=time){
                low = mid;
            } else {
                high = mid;
            }
            
        }
        if(closestEntry==null){
            closestEntry = leadChangeList.get(high).get(0)<=time ? leadChangeList.get(high) : leadChangeList.get(low);
        }
        return closestEntry;
        
    }
    
    
    public TopVotedCandidate(int[] persons, int[] times) {
        leadChangeList = new ArrayList<ArrayList<Integer>>();
        
        HashMap<Integer,Integer> voteCount = new HashMap<Integer,Integer>();
        Integer currLeader = null;
        for(int i=0;i<times.length;++i){
            int newVoteCount =voteCount.getOrDefault(persons[i],0)+1;
            voteCount.put(persons[i],newVoteCount);
            if(currLeader==null || (currLeader!=persons[i] && voteCount.get(currLeader)<=newVoteCount)){
                currLeader=persons[i];
                leadChangeList.add(new ArrayList<Integer>(Arrays.asList(times[i],persons[i])));
            }
            
        }
        
        // System.out.println(leadChangeList);
        
        
    }
    
    public int q(int t) {
        return getClosestLeadEntryForTime(t).get(1);
    }
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj.q(t);
 */

File: leetcode/algorithms/binary-search/online-election.ts
class TopVotedCandidate {
    times: number[];
    persons: number[];
    leader: number[];
    constructor(persons: number[], times: number[]) {
        this.persons = persons;
        this.times = times;
        this.leader = [];
        let currLead: number = persons[0];
        const voteCount: Map<number,number> = new Map<number,number>();
        for(let i = 0;i<times.length;++i){
                
            voteCount.set(persons[i],(voteCount.get(persons[i]) || 0)+1);
                if(voteCount.get(persons[i])>=voteCount.get(currLead)){
                    currLead = persons[i];
                }
            
            this.leader.push(currLead);
        }
    }

    q(t: number): number {
        let h: number = this.times.length-1,l: number=0;
        let targetIndex: number | null = null;
        while(h-l>1){
            const m = l+Math.floor((h-l)/2);
            if(this.times[m]===t){
                targetIndex=m;
                break;
            } else if(this.times[m]>t){
                h=m;
            } else {
                l=m;
            }
        }
        if(targetIndex===null){
            if(this.times[h]<=t){
                targetIndex = h;
            } else {
                targetIndex = l;
            }
        }
        return this.leader[targetIndex]
    }
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * var obj = new TopVotedCandidate(persons, times)
 * var param_1 = obj.q(t)
 */